### ISSUE TITLE
Undo feature

### ISSUE THREAD

**Issue by FredrikNoren** [FredrikNoren/ungit#124]
A way to undo changes, perhaps with a history of what you've done.

**Comment by codingtwinky** [comment]
About one year later...

What is the scope of this?  Some operation are easy to undo, but some operation we would require something more elaborate, such as undoing "disregard" operation.

**Comment by FredrikNoren** [comment]
Yeah I wonder though if discard could be reimplemented as a commit + move
branch back one commit. That way you could actually use the ref pointer to
get the data back

**Comment by Ajedi32** [comment]
Yeah, I'd prefer it if all operations were made to be reversible. As a starting point though, a front end to `git reflog` would suffice, maybe with a undo button that pops up when you delete or reset a branch in a way that causes commits to become inaccessible via regular branch pointers. For destructive operations where undo isn't yet implemented, or can't be implemented, a confirmation dialog should be shown.

**Comment by codingtwinky** [comment]
I do like the idea of having a branch that helps us with undoing disregards.  How about for other operations?  As far as I know there is no git undo.  So what would be the best way to do this?  Maintain a Map of git operation to another operations that undo?

**Comment by Ajedi32** [comment]
@codingtwinky Well, fundamentally there's actually not a very large set of operations git can do. Basically, it can move, create, and delete branch pointers (local and remote), add and remove text from the staging area, and update the working directory with data from files in git. I think that's it really. Am I missing anything?

The first part (manipulating branch pointers) should be very easy to add undo functionality for, since git keeps track of all the previous positions of your branches by default anyway (reflog). The only thing I don't think it tracks is when you delete branches, though I'm not sure on that point.

Manipulation of the staging area would be a bit harder to add undo functionality to, since git doesn't already keep track of operations performed there. It does add files in the staging area to its database though I believe, so all the previous states of the files in the staging area should be tracked, we just have to hook into that functionality and remember to keep track of operations which modify the staging area so we can undo them.

As for modifications of files in the working directory, that's a bit harder. Git currently doesn't store any information on untracked or unstaged files, so if we want to add undo functionality there, we'd have to force git to save "dummy commits" for files in the working directory every time it modifies them. This might be undesirable though in cases where there are large, untracked files in the working directory, so we'd also have to deal with that problem when implementing this bit.

Personally, I say we could just start with allowing undoing branch pointer manipulations. Allowing undos on changes to the staging area and working directory would come later.

**Comment by TijmenW** [comment]
Maybe it is possible to use `git reflog`?

**Comment by antoninadert** [comment]
Apparently Gitup did this but this is only for Mac. Are they people knowledgeable about this around ?

**Comment by wmertens** [comment]
GitUp does this by storing refs and a message before any change, and then when you undo, it restores the refs. It's a nicer way of spelunking the reflog.
